Documentation: WAF: New lease explosions overview (#24683)
hashicorp/vault: website/content/docs/concepts/lease-explosions.mdx

   max_leases=100 \
   path="auth/approle" \
   role="webapp"
```

Example output:

<CodeBlockConfig hideClipboard>

```plaintext
Success! Data written to: sys/quotas/lease-count/webapp-tokens
```

</CodeBlockConfig>

Confirm the setting.

```shell-session
$ vault read sys/quotas/lease-count/webapp-tokens
```

Example output:

<CodeBlockConfig hideClipboard>

```plaintext
Key            Value
---            -----
counter        0
inheritable    true
max_leases     100
name           webapp-tokens
path           auth/approle/
role           webapp
type           lease-count
```

</CodeBlockConfig>

</Tab>

</Tabs>

The limit is set to 100 leases for the AppRole auth method role named webapp.

<Note>

Enabling the rate limit audit logging may have an impact on the Vault performance if the volume of rejected requests is large.

</Note>

Review these resources for a deeper dive into controlling Vault resources:

- [Vault resource quotas](/vault/docs/concepts/resource-quotas)

- [Vault Enterprise lease count quotas](/vault/docs/enterprise/lease-count-quotas)

- [Query audit device logs](/vault/tutorials/monitoring/query-audit-device-logs)

#### Token type

> In some use cases, batch tokens can be a better fit for a use case than service tokens with respect to lease explosion. Review the following resources for help deciding when to use batch tokens and when to use service tokens:

Reviewer: ```suggestion
In some use cases, batch tokens can be a better fit than service tokens with respect to lease explosion. Review the following resources for help deciding when to use batch tokens and when to use service tokens:
```

Should we give an example of what would make them a better or worse fit? How should the reader decide?

- [Vault service tokens vs batch tokens](/vault/tutorials/tokens/batch-tokens#service-tokens-vs-batch-tokens)

- [Service vs batch token lease handling](/vault/docs/concepts/tokens#service-vs-batch-token-lease-handling)

## Mitigating lease explosions

> Ultimately, the number of leases a system can handle is unique to the Vault deployment and environment. 

Reviewer: Should we give a tiny bit more intro here. Maybe something like "The mitigation strategy you use will depend on what caused the explosion."? It also might be a good idea to be explicit about where users should start if they're trying to mitigate. For example, should they try these strategies in order or select the one that matches their problem? If they are supposed to choose one it might be nice to give them some symptoms to look for in order to choose, or a way to check if a strategy listed here is right for them.

### Increase resources

Increasing available resources in your Vault cluster can help mitigate lease explosion and allow for cluster recovery. Review [hardware sizing](/well-architected-framework/zero-trust-security/raft-reference-architecture#hardware-sizing-for-vault-servers), and focus on increasing available RAM.

#### Within Vault

Use the information from the Implementation guardrails section to adjust TTL values from the default values according to your use case needs.

#### External to Vault

You can use firewalls or load balancers to limit API calls to Vault from aberrant clients.)

[Knowledge base article around load balancing](https://support.hashicorp.com/hc/en-us/articles/14496042865427-Vault-Global-Load-Balancing-Patterns)
[Vault & load balancing](/vault/tutorials/day-one-raft/raft-reference-architecture#load-balancer-recommendations)

## Cleaning up environment after lease explosions

> Once the acute event subsides, the Vault active node will continue to purge leases. Sometimes, the explosion is so great, manual intervention is needed to revoke [leases](/vault/api-docs/system/leases). If you are running a version of Vault prior to 1.13.0, this lease revocation can cause further performance degradation.

Reviewer: ```suggestion
Once the acute event subsides, the Vault active node will continue to purge leases. Sometimes, the explosion is so great, you will need to manually intervene to revoke [leases](/vault/api-docs/system/leases). If you are running a version of Vault prior to 1.13.0, this lease revocation can cause further performance degradation.
```

> Revoking or forcefully revoking leases is potentially a dangerous operation. You should ensure that you have recent valid snapshots of the cluster. Users of Vault versions prior to 1.13.0 on integrated storage must also perform freelist compaction. Vault Enterprise customers should consider proactively contacting the [Customer Success team](https://support.hashicorp.com) for assistance with this process.

Reviewer: ```suggestion
Revoking or forcefully revoking leases is potentially a dangerous operation. You should ensure that you have recent valid snapshots of the cluster. Users of Vault versions prior to 1.13.0 on integrated storage must also perform freelist compaction. Vault Enterprise customers should consider proactively contacting the [Customer Success team](https://support.hashicorp.com) for help with this process.
```
